Source/DafnyCore/AST/Tokens.cs

```diff
@@ -314,15 +314,17 @@ public class BoogieRangeToken : TokenWrapper {
   public IToken StartToken { get; }
   public IToken EndToken { get; }
   public IToken NameToken { get; }
+  public IToken UniquelyIdentifyingToken { get; }
```

Comment:
FYI, I tried reusing NameToken for the purpose above but it broke 150 test as apparently it changes the way related locations are reported, and I could not pinpoint exactly how this was done. I'm not sure exactly why there is a NameToken in a BoogieRange on the first place. Could you please explain me why?

In your interface, in any case, because of my observation, we would have to make sure we make the distinction between whether `NameOrKeyword` is a name or a keyword. However, to make that distinction, we would then need one more boolean or enum field. So what's the point of storing one token that can contain two different pieces of information + a boolean that helps knowing what this token is? Plus, couldn't we have a case where Name is required for its own purpose (which I don't know of), and UniquelyIdentifyingToken is used for its own purpose, and they are not null at the same time?

Reply:
> In your interface, in any case, because of my observation, we would have to make sure we make the distinction between whether NameOrKeyword is a name or a keyword

What observation?

> FYI, I tried reusing NameToken for the purpose above but it broke 150 test as apparently it changes the way related locations are reported, and I could not pinpoint exactly how this was done. I'm not sure exactly why there is a NameToken in a BoogieRange on the first place. Could you please explain me why?

It has the same purpose as `UniquelyIdentifyingToken`, to report on specific positions.

Reply:
> > FYI, I tried reusing NameToken for the purpose above but it broke 150 test as apparently it changes the way related locations are reported, and I could not pinpoint exactly how this was done. I'm not sure exactly why there is a NameToken in a BoogieRange on the first place. Could you please explain me why?

That's my observation. NameToken was changing the way reports were done (error messages changing from "`Could not prove: b > c` to `Related location:` for example) and I am not sure how, but given the fact that it's called `NameToken`, is undocumented, and from the name is not meant to uniquely identify an expression. it would have helped me to understand what NameToken exactly is.

> It has the same purpose as UniquelyIdentifyingToken, to report on specific positions.

What positions? Can you please give me an example of how it is used?

Reply:
I agree with you the current code is confusing. I've tried to improve it here: https://github.com/dafny-lang/dafny/pull/4630